### PR TITLE
plpgsql: implement RAISE statements

### LINKED ISSUE #105251: plpgsql: add support for raising errors and messages
Postgres PLpgSQL allows the function to use a `RAISE` statement to send messages back to the client and/or throw an error: https://www.postgresql.org/docs/current/plpgsql-errors-and-messages.html#PLPGSQL-STATEMENTS-RAISE
Messages are returned to the client asynchronously after the `RAISE` while execution continues, while a thrown error stops execution and propagates up to the exception block (if any).

We could potentially define builtin functions that have the desired side-effects when called, though we'll have to be careful about making sure the optimizer doesn't re-order/eliminate the side-effects. We may have to introduce some optimization barrier to make sure the side-effect happens at the expected place, e.g. a `RAISE MESSAGE` after some operation completes shouldn't be executed until after that operation is done.

Epic: CRDB-799

Jira issue: CRDB-28940

### PR DESCRIPTION
**Pull request by DrewKimball**
#### builtins: add builtin function to raise a notice synchronously to the client

This patch adds a builtin function, `crdb_internal.plpgsql_raise`, which allows
the caller to send a notice to the client with specified severity, message,
detail, hint, and PG code. The notice is immediately flushed to the client
instead of being buffered until the query result is closed. This functionality
will be used to implement the PLpgSQL `RAISE` statement.

The `crdb_internal.plpgsql_raise` builtin is undocumented and intended only
for internal use.

#### plpgsql: implement parsing for RAISE statements

This patch adds parser support for PLpgSQL `RAISE` statements. This ncludes
all syntax forms apart from the empty `RAISE`, which is only valid in
combination with (currently unimplemented) `EXCEPTION` blocks. A future
commit will add support in the optbuilder as well.

#### plpgsql: implement RAISE statement

This patch adds support for the PLpgSQL `RAISE` statement. The `RAISE`
statement can send messages back to the client during execution, as well
as raise a user-specified error. There are a few variations on the syntax,
but in general `RAISE` statements have a log level (default `EXCEPTION`),
a message (if not specified, the code string is used), and various options:
`DETAIL`, `HINT`, `ERRCODE` etc.

With log level `EXCEPTION` the error is returned just like any other error,
but for other levels it is sent as a notice to the client and flushed
synchronously before execution continues. This feature is often used to
track progress, since the notices are sent before execution finishes.

Fixes #105251

Release note (sql change): Added support for the PLpgSQL `RAISE` statement,
which allows sending notices to the client and raising errors. Currently the
notice is only sent to the client; support for logging notices is left for
future work.

### CODE REVIEW

**Review by rharding6373 — CHANGES_REQUESTED** [review]
Impressive work!

> Reviewed 25 of 25 files at r1, 4 of 4 files at r2, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-:-N_G3LrG-PpGUd1TTbMS:b-tgv6b6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @DrewKimball, and @mgartner)

___
*[`-- commits` line 20 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FL4fvBL7t3XTVymoQ):*
s/ncludes/includes
___
*[`pkg/sql/conn_io.go` line 1087 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FBquB-_AfcRI6MMnF:-N_FBquB-_AfcRI6MMnG:bp8oi7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/9025d5348896146a105c5c6267670414f695ca6f/pkg/sql/conn_io.go#L1087)):*
> ```Go
> // SendNotice is part of the RestrictedCommandResult interface.
> func (r *streamingCommandResult) SendNotice(ctx context.Context, notice pgnotice.Notice) error {
> 	// Unimplemented: the internal executor does not support notices.
> ```

nit: indentation.

Also, if this is unimplemented or unsupported, should we return an unimplemented error?
___
*[`pkg/sql/notice.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FDxvPBFij5iyV9AYt:-N_FDxvPBFij5iyV9AYu:bpryf2h) ([raw file](https://github.com/cockroachdb/cockroach/blob/9025d5348896146a105c5c6267670414f695ca6f/pkg/sql/notice.go#L55)):*
> ```Go
> func (p *planner) SendClientNotice(ctx context.Context, notice pgnotice.Notice) error {
> 	if log.V(2) {
> 		log.Infof(ctx, "sending notice: %+v", notice)
> ```

Would it be useful to include the client we're sending to as part of the logging?
___
*[`pkg/sql/opt/optbuilder/plpgsql.go` line 462 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FsRPW10_6cb5xsrfx:-N_FsRPW10_6cb5xsrfy:b6nhcxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/8821ba0f4da6cbdb82fe3114546478b668e500c7/pkg/sql/opt/optbuilder/plpgsql.go#L462)):*
> ```Go
> 	buildOptionExpr := func(name string, expr plpgsqltree.PLpgSQLExpr, isDup bool) opt.ScalarExpr {
> 		if isDup {
> 			panic(pgerror.Newf(pgcode.PLpgSQL, "RAISE option already specified: %s", name))
> ```

nit: in postgres, dupes use the syntax error code:

```
rharding=# create function f() returns void language plpgsql as $$ 

  begin raise notice using message = 'hello', detail='world', detail='again'; end $$;

CREATE FUNCTION

rharding=# select f();

ERROR:  42601: RAISE option already specified: DETAIL

CONTEXT:  PL/pgSQL function f() line 2 at RAISE

LOCATION:  exec\_stmt\_raise, pl\_exec.c:3812
```
___
*[`pkg/sql/opt/optbuilder/plpgsql.go` line 544 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_Fv7Sz-aRH84D7N3US:-N_Fudg3E91wiIOgAu1K:b-9a90du) ([raw file](https://github.com/cockroachdb/cockroach/blob/8821ba0f4da6cbdb82fe3114546478b668e500c7/pkg/sql/opt/optbuilder/plpgsql.go#L544)):*
> ```Go
> 	}
> 	if argIdx < len(args) {
> 		panic(pgerror.Newf(pgcode.PLpgSQL, "too many parameters specified for RAISE"))
> ```

In postgres this error happens during function creation, but IIUC this error happens when the function with the raise is called. Could you please clarify?
___
*[`pkg/sql/plpgsql/parser/plpgsql.y` line 1118 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FRC-34ahx9YAFeSRR:-N_FRC-34ahx9YAFeSRS:b1i731d) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0161776ba896368b121084be75046f39fd0426/pkg/sql/plpgsql/parser/plpgsql.y#L1118)):*
> ```Yacc
> 
> option_expr:
>   USING option_type assign_operator
> ```

I don't think this will correctly parse multiple options correctly. In postgres, you specify multiple options like this:

```
create function f() returns void language plpgsql as $$ 
  begin 
    raise notice using message = 'hello', detail='world'; 
  end $$;
```

But it looks like the parser currently looks for:
```
raise notice using message = 'hello', using detail='world';
```

This would be incorrect and doesn't parse in postgres.

[Coming back to this after reviewing everything]: I saw in a later TODO that you noticed this too but it seemed allowed by the syntax. I sympathize that specifications like`RAISE \[ _`level`_ \] SQLSTATE '_`sqlstate`_' \[ USING _`option`_ = _`expression`_ \[, ... \] \];` in https://www.postgresql.org/docs/current/plpgsql-errors-and-messages.html are hard to parse. Ultimately I think that what postgres is doing is consistent with the spec and should be the source of truth for the syntax.
___
*[`pkg/sql/plpgsql/parser/plpgsql.y` line 1046 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_G0wGy1-extXF_uYNI:-N_G0fx9CraNDjUEbytG:bfgzx76) ([raw file](https://github.com/cockroachdb/cockroach/blob/8821ba0f4da6cbdb82fe3114546478b668e500c7/pkg/sql/plpgsql/parser/plpgsql.y#L1046)):*
> ```Yacc
>     return unimplemented(plpgsqllex, "empty RAISE statement")
>   }
> | RAISE error_level SCONST optional_format_exprs optional_option_exprs ';'
> ```

minor nit: we seem to default to using "opt_" instead of "optional_" in names in this file.
___
*[`pkg/sql/plpgsql/parser/plpgsql.y` line 1080 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_G17g7BGDIyZ6f370i:-N_G17g84g6XrYqQqEW2:bbuou0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/8821ba0f4da6cbdb82fe3114546478b668e500c7/pkg/sql/plpgsql/parser/plpgsql.y#L1080)):*
> ```Yacc
> ;
> 
> error_level:
> ```

nit: since error_level is optional, I think this should be renamed.
___
*[`pkg/sql/sem/builtins/builtins.go` line 8051 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FJEoTDAKZ6rVqTFQK:-N_FJEoTDAKZ6rVqTFQL:b-bst6lm) ([raw file](https://github.com/cockroachdb/cockroach/blob/9025d5348896146a105c5c6267670414f695ca6f/pkg/sql/sem/builtins/builtins.go#L8051)):*
> ```Go
> 					var code string
> 					if regexp.MustCompile(`[A-Z0-9]{5}`).MatchString(codeString) {
> 						// The supplied argument is a valid PG code.
> ```

Isn't it only a valid PG code if it's one of the codes mapped to in PLpgSQLConditionNameToCode or in https://www.postgresql.org/docs/current/errcodes-appendix.html?
___
*[`pkg/sql/logictest/testdata/logic_test/udf_plpgsql` line 571 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_G1Wkg-jDC3jHIlSWg:-N_G1Wkg-jDC3jHIlSWh:bt8pfzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8821ba0f4da6cbdb82fe3114546478b668e500c7/pkg/sql/logictest/testdata/logic_test/udf_plpgsql#L571)):*
> ```
> CREATE OR REPLACE FUNCTION f() RETURNS INT AS $$
>   BEGIN
>     RAISE DEBUG 'foo';
> ```

Could you add a test without a level, e.g. `RAISE 'foo'`?
___
*[`pkg/sql/plpgsql/parser/testdata/stmt_raise` line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FdSVf2LyCmqxsoF0w:-N_FdSVf2LyCmqxsoF0x:b-txa5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0161776ba896368b121084be75046f39fd0426/pkg/sql/plpgsql/parser/testdata/stmt_raise#L36)):*
> ```
> DECLARE
> BEGIN
>   RAISE LOG 'Nonexistent ID --> %', user_id;
> ```

Also add a test that parses multiple format expressions.
___
*[`pkg/sql/plpgsql/parser/testdata/stmt_raise` line 122 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FbvXX2ll26myLSzDO:-N_FbvXX2ll26myLSzDP:b-ssr7gi) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0161776ba896368b121084be75046f39fd0426/pkg/sql/plpgsql/parser/testdata/stmt_raise#L122)):*
> ```
>   RAISE internal_screaming
>   USING MESSAGE = 'blah blah blah',
>   USING COLUMN = 'foo';
> ```

As mentioned in a previous comment, this doesn't parse in postgres, so these parser tests need an update.

**Comment by knz** [comment]
I'm curious, why does the built-in not interrupt execution by returning an error when the severity "error" is specified?

**Comment by DrewKimball** [comment]
>I'm curious, why does the built-in not interrupt execution by returning an error when the severity "error" is specified?

It should, what makes you think it doesn't?

**Comment by knz** [comment]
> It should, what makes you think it doesn't?

Your own PR description:

> With log level EXCEPTION the error is returned just like any other error, but for other levels it is sent as a notice to the client and flushed synchronously before execution continues.

If ERROR stops execution too, the PR description should mention that.

**Comment by DrewKimball** [comment]
>>Your own PR description:
>
>With log level EXCEPTION the error is returned just like any other error, but for other levels it is sent as a notice to the client and flushed synchronously before execution continues.
>
>If ERROR stops execution too, the PR description should mention that.

I see, I'll make sure to clarify things, thanks.

**Review by chengxiong-ruan — APPROVED** [review]
The parser change LGTM. nice work.

**Review by DrewKimball — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106351)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rharding6373)

___
*[`-- commits` line 20 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FL4fvBL7t3XTVymoQ):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

s/ncludes/includes
</blockquote></details>

Done.
___
*[`pkg/sql/conn_io.go` line 1087 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FBquB-_AfcRI6MMnF:-N_jvRhfCJPfsOq9UPoi:b-epoiuk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9025d5348896146a105c5c6267670414f695ca6f/pkg/sql/conn_io.go#L1087)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

nit: indentation.

Also, if this is unimplemented or unsupported, should we return an unimplemented error?
</blockquote></details>

I don't think we expect to call this method from the internal executor.
___
*[`pkg/sql/notice.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FDxvPBFij5iyV9AYt:-N_jvd5mBZzpotDsj9pu:b-3lar5g) ([raw file](https://github.com/cockroachdb/cockroach/blob/9025d5348896146a105c5c6267670414f695ca6f/pkg/sql/notice.go#L55)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Would it be useful to include the client we're sending to as part of the logging?
</blockquote></details>

What information about the client do you mean?
___
*[`pkg/sql/opt/optbuilder/plpgsql.go` line 462 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FsRPW10_6cb5xsrfx:-N_jvsp85qzLZabqE5et:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8821ba0f4da6cbdb82fe3114546478b668e500c7/pkg/sql/opt/optbuilder/plpgsql.go#L462)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

nit: in postgres, dupes use the syntax error code:

```
rharding=# create function f() returns void language plpgsql as $$ 

  begin raise notice using message = 'hello', detail='world', detail='again'; end $$;

CREATE FUNCTION

rharding=# select f();

ERROR:  42601: RAISE option already specified: DETAIL

CONTEXT:  PL/pgSQL function f() line 2 at RAISE

LOCATION:  exec\_stmt\_raise, pl\_exec.c:3812
```
</blockquote></details>

Done.
___
*[`pkg/sql/opt/optbuilder/plpgsql.go` line 544 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_Fv7Sz-aRH84D7N3US:-N_jwEUUEoXn6nszBY9g:box24jv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8821ba0f4da6cbdb82fe3114546478b668e500c7/pkg/sql/opt/optbuilder/plpgsql.go#L544)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

In postgres this error happens during function creation, but IIUC this error happens when the function with the raise is called. Could you please clarify?
</blockquote></details>

Once the todo [here](https://github.com/cockroachdb/cockroach/blob/adbcda5e1a08fbee64cc56d338c36a504c0fabc9/pkg/sql/opt/optbuilder/create_function.go#L229-L231) is addressed, the error will happen upon creation, as well as other errors that are thrown here.
___
*[`pkg/sql/plpgsql/parser/plpgsql.y` line 1118 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FRC-34ahx9YAFeSRR:-N_k9UltBPs0f5bkpnOl:briqu4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0161776ba896368b121084be75046f39fd0426/pkg/sql/plpgsql/parser/plpgsql.y#L1118)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

I don't think this will correctly parse multiple options correctly. In postgres, you specify multiple options like this:

```
create function f() returns void language plpgsql as $$ 
  begin 
    raise notice using message = 'hello', detail='world'; 
  end $$;
```

But it looks like the parser currently looks for:
```
raise notice using message = 'hello', using detail='world';
```

This would be incorrect and doesn't parse in postgres.

[Coming back to this after reviewing everything]: I saw in a later TODO that you noticed this too but it seemed allowed by the syntax. I sympathize that specifications like`RAISE \[ _`level`_ \] SQLSTATE '_`sqlstate`_' \[ USING _`option`_ = _`expression`_ \[, ... \] \];` in https://www.postgresql.org/docs/current/plpgsql-errors-and-messages.html are hard to parse. Ultimately I think that what postgres is doing is consistent with the spec and should be the source of truth for the syntax.
</blockquote></details>

Huh, I actually just didn't realize the `USING` was omitted for subsequent options. Thanks for pointing that out.
___
*[`pkg/sql/plpgsql/parser/plpgsql.y` line 1046 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_G0wGy1-extXF_uYNI:-N_k8stHBaxLjqKhyQTZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8821ba0f4da6cbdb82fe3114546478b668e500c7/pkg/sql/plpgsql/parser/plpgsql.y#L1046)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

minor nit: we seem to default to using "opt_" instead of "optional_" in names in this file.
</blockquote></details>

Done.
___
*[`pkg/sql/plpgsql/parser/plpgsql.y` line 1080 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_G17g7BGDIyZ6f370i:-N_k9C_w1-9atpOAQKLa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8821ba0f4da6cbdb82fe3114546478b668e500c7/pkg/sql/plpgsql/parser/plpgsql.y#L1080)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

nit: since error_level is optional, I think this should be renamed.
</blockquote></details>

Done.
___
*[`pkg/sql/sem/builtins/builtins.go` line 8051 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FJEoTDAKZ6rVqTFQK:-N_jyAcM5QP3ZstK_z8N:bq0c8sa) ([raw file](https://github.com/cockroachdb/cockroach/blob/9025d5348896146a105c5c6267670414f695ca6f/pkg/sql/sem/builtins/builtins.go#L8051)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Isn't it only a valid PG code if it's one of the codes mapped to in PLpgSQLConditionNameToCode or in https://www.postgresql.org/docs/current/errcodes-appendix.html?
</blockquote></details>

It's allowed to be anything: https://www.postgresql.org/docs/current/plpgsql-errors-and-messages.html
The docs say it can't be `00000`, but a quick test shows that works too:
```
postgres=# create function f() returns int language plpgsql as $$ begin raise sqlstate '00000'; return 0; end $$;
CREATE FUNCTION
postgres=# select f();
ERROR:  00000
CONTEXT:  PL/pgSQL function f() line 1 at RAISE
```
___
*[`pkg/sql/logictest/testdata/logic_test/udf_plpgsql` line 571 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_G1Wkg-jDC3jHIlSWg:-N_jz1oL2YoopSa_A-1y:bvuv72w) ([raw file](https://github.com/cockroachdb/cockroach/blob/8821ba0f4da6cbdb82fe3114546478b668e500c7/pkg/sql/logictest/testdata/logic_test/udf_plpgsql#L571)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Could you add a test without a level, e.g. `RAISE 'foo'`?
</blockquote></details>

Hm, could've sworn I did that. Done. Caught a bug, too - I incorrectly set the pgcode in that case before.
___
*[`pkg/sql/plpgsql/parser/testdata/stmt_raise` line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FdSVf2LyCmqxsoF0w:-N_kHyq64x-CtHs1qaMK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0161776ba896368b121084be75046f39fd0426/pkg/sql/plpgsql/parser/testdata/stmt_raise#L36)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Also add a test that parses multiple format expressions.
</blockquote></details>

Done.
___
*[`pkg/sql/plpgsql/parser/testdata/stmt_raise` line 122 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FbvXX2ll26myLSzDO:-N_kI05EEstbij1GA_PA:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a0161776ba896368b121084be75046f39fd0426/pkg/sql/plpgsql/parser/testdata/stmt_raise#L122)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

As mentioned in a previous comment, this doesn't parse in postgres, so these parser tests need an update.
</blockquote></details>

Fixed.

**Review by rharding6373 — APPROVED** [review]
:lgtm: Nice work, Drew!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-:-N_kT5Tb-bMxWZyuyQ1x:b-k8akd4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/notice.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FDxvPBFij5iyV9AYt:-N_kRgLi3pI6UduXRmke:b7hjt3m) ([raw file](https://github.com/cockroachdb/cockroach/blob/9025d5348896146a105c5c6267670414f695ca6f/pkg/sql/notice.go#L55)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

What information about the client do you mean?
</blockquote></details>

Something like "sending notice to client [ID]", but on second thought I don't think it would be super useful even if it's possible.
___
*[`pkg/sql/opt/optbuilder/plpgsql.go` line 544 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_Fv7Sz-aRH84D7N3US:-N_kSIHj8NONPU4M7r-n:bcr5s31) ([raw file](https://github.com/cockroachdb/cockroach/blob/8821ba0f4da6cbdb82fe3114546478b668e500c7/pkg/sql/opt/optbuilder/plpgsql.go#L544)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Once the todo [here](https://github.com/cockroachdb/cockroach/blob/adbcda5e1a08fbee64cc56d338c36a504c0fabc9/pkg/sql/opt/optbuilder/create_function.go#L229-L231) is addressed, the error will happen upon creation, as well as other errors that are thrown here.
</blockquote></details>

Got it, thanks.
___
*[`pkg/sql/sem/builtins/builtins.go` line 8051 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106351#-N_FJEoTDAKZ6rVqTFQK:-N_kSngrCqCOnHbpQkUM:b2nhoia) ([raw file](https://github.com/cockroachdb/cockroach/blob/9025d5348896146a105c5c6267670414f695ca6f/pkg/sql/sem/builtins/builtins.go#L8051)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

It's allowed to be anything: https://www.postgresql.org/docs/current/plpgsql-errors-and-messages.html
The docs say it can't be `00000`, but a quick test shows that works too:
```
postgres=# create function f() returns int language plpgsql as $$ begin raise sqlstate '00000'; return 0; end $$;
CREATE FUNCTION
postgres=# select f();
ERROR:  00000
CONTEXT:  PL/pgSQL function f() line 1 at RAISE
```
</blockquote></details>

TIL

**Comment by DrewKimball** [comment]
TFTRs!
